test/mpi/comm/cmsplit_type.c

```diff
@@ -55,6 +55,67 @@ int main(int argc, char *argv[])
         MPI_Comm_free(&comm);
     }
 
+    /* Check to see if MPI_COMM_TYPE_HW_GUIDED works correctly */
+    for (i = 0; split_topo[i]; i++) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[i]);
+        MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+        if (comm != MPI_COMM_NULL) {
+            int newsize;
+            MPI_Comm_size(comm, &newsize);
+            if (newsize > size) {
+                printf("Expected comm size to be at most the node size\n");
+                errs++;
+            }
+            MPI_Comm_free(&comm);
+        }
+        MPI_Info_free(&info);
+    }
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass different info values, it must return
+     * MPI_COMM_NULL.*/
+    MPI_Info_create(&info);
+    if (rank % 2 == 0) {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[0]);
+    } else {
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[1]);
+    }
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("Expected MPI_COMM_NULL due to different info values.\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    MPI_Info_free(&info);
+
+    /* Test MPI_COMM_TYPE_HW_GUIDED: pass info to some processes, it must return
+     * MPI_COMM_NULL.*/
+    if (rank % 2 == 0) {
+        MPI_Info_create(&info);
+        MPI_Info_set(info, "mpi_hw_resource_type", split_topo[0]);
+    } else
+        info = MPI_INFO_NULL;
+    MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_HW_GUIDED, 0, info, &comm);
+    if (comm != MPI_COMM_NULL) {
+        printf("Expected MPI_COMM_NULL due to inconsistent infohints.\n");
+        errs++;
+        MPI_Comm_free(&comm);
+    }
+    if (rank % 2 == 0)
+        MPI_Info_free(&info);
```

Comment:
new test added

Reply:
It looks like the standard allows one process to specify info but the others pass `INFO_NULL`. `COMM_NULL` is returned **only for the processes with `INFO_NULL`** (see quoted standard below). Thus, I assume that a valid newcomm should be returned for the processes passed a valid info.

>If an MPI process does not meet the above criteria, then MPI_COMM_NULL is returned in newcomm for such process.
MPI_COMM_NULL is also returned in newcomm in the following cases:
• No info key is provided.
• The info handle does not include the key "mpi_hw_resource_type" .

Besides, the tests covers only the INFO_NULL case, maybe we want to also add the cases for "info without key" and "info without mpi_hw_resource_type key"?